rewindable-index/test/RewindableIndex/Spec/Sqlite.hs

```diff
@@ -0,0 +1,289 @@
+
+  queryStorage qi memoryEs (Handle h) (QEvents f)  = do
+    Sql.execute_ h "BEGIN"
+    aggregate <- queryStorage qi memoryEs (Handle h) QAccumulator
+    es' :: [StorableEvent Handle] <-
+      case qi of
+        QEverything -> Sql.query_ h "SELECT * from index_property_cache ORDER BY point ASC"
+        QInterval _ end ->
+          Sql.query h "SELECT * from index_property_cache WHERE point <= ? ORDER BY point ASC" (Sql.Only end)
+    Sql.execute_ h "COMMIT"
+    let es'' = filterWithQueryInterval qi (es' ++ toList memoryEs)
+    pure $ foldl' ((fst .) . indexedFn f) aggregate es''
+
+instance Rewindable Handle IO where
+  rewindStorage :: StorablePoint Handle -> Handle -> IO (Maybe Handle)
+  rewindStorage pt (Handle h) = do
+    Sql.execute h "DELETE FROM index_property_cache WHERE point > ?" (Sql.Only pt)
+    pure . Just $ Handle h
+
+instance Resumable Handle IO where
+  resumeFromStorage :: Handle -> IO [StorablePoint Handle]
+  resumeFromStorage (Handle h) = do
+    es' :: [StorableEvent Handle] <-
+      Sql.query_ h "SELECT * FROM index_property_cache ORDER BY point DEC"
+    pure $ fmap sePoint es'
+
+-- * Conversions
+
+type IndexT = Index Int Int ()
+
+conversion
+  :: Conversion (PropertyM IO) Int Int ()
+conversion = Conversion
+  { cView = getView
+  , cHistory = getHistory
+  , cNotifications = getNotifications
+  , cMonadic = monadic
+  }
+
+getNotifications
+  :: IndexT
+  -> PropertyM IO [()]
+getNotifications _ = pure []
+
+getHistory
+  :: IndexT
+  -> PropertyM IO (Maybe [Int])
+getHistory ix = do
+  mix <- run ix
+  case mix of
+    Nothing       -> pure Nothing
+    Just (ix', _) -> liftIO $ do
+      let st = ix' ^. state
+          -- f  = ix' ^. fn
+          f  = getFunction ix
+          sz = ix' ^. state . config . memoryBufferSize -- + 1
+      es <- Storable.getEvents st
+      let qs = map (\p -> QInterval p p)
+             $ map sePoint es
+      rs  <- fmap getResult <$> mapM (\qi -> Storable.query qi st (QEvents f)) qs
+      if null rs
+      then do
+        Result ag0 <- Storable.query QEverything st QAccumulator
+        pure $ Just [ag0]
+      else
+        pure . Just . take sz $ reverse rs -- ++ [ag0]
+
+getView
+  :: IndexT
+  -> PropertyM IO (Maybe (IndexView Int))
+getView ix = do
+  mix <- run ix
+  case mix of
+    Nothing       -> pure Nothing
+    Just (ix', _) -> do
+      let maxSize = ix' ^. state . config . memoryBufferSize + 1
+      es <- liftIO $ Storable.getEvents (ix' ^. state)
+      Just rs <- getHistory ix
+      let sz = if null es then 1 else length es + 1
+      -- let sz = if length rs == 1 then 1 else length rs + 1
+      pure . Just $
+        IndexView { ixDepth = maxSize
+                  , ixView  = head rs
+                  , ixSize  = min maxSize sz
+                  }
+
+monadic
+  :: PropertyM IO Property
+  -> Property
+monadic = monadicIO
+
+lookupPoint
+  :: Int
+  -> Config
+  -> MaybeT IO (StorablePoint Handle)
+-- TODO: this needs a lot of comments
```

Comment:
yes, plz as I am having hard time following the logic here

Reply:
Done.